File: 6_tree/线段树/715. Range 模块-动态开点线段树.py
# 1 <= left < right <= 1e9

# 由于值域很大且是在线query，无法做离散化
# 所以需要动态开点
# 区间操作，线段树


from typing import Optional


class Node:
    # __slots__允许我们声明并限定类成员，并拒绝类创建__dict__和__weakref__属性以节约内存空间。
    __slots__ = ('isTracked', 'lazy', 'left', 'right')

    def __init__(
        self,
        isTracked=False,  # 表示当前区间所有元素是否被追踪
        lazy=False,
        left: Optional['Node'] = None,
        right: Optional['Node'] = None,
    ) -> None:
        self.isTracked = isTracked
        self.lazy = lazy
        self.left = left
        self.right = right


class SegmentTree:
    def __init__(self) -> None:
        self._root = Node()

    def update(self, left: int, right: int, delta: bool) -> None:
        self._update(left, right, 0, int(1e9 + 10), self._root, delta)

    def query(self, left: int, right: int) -> bool:
        return self._query(left, right, 0, int(1e9 + 10), self._root)

    def _update(self, L: int, R: int, l: int, r: int, root: Node, delta: bool) -> None:
        if L <= l <= r <= R:
            root.isTracked = delta
            root.lazy = True
            return

        self._pushDown(root)
        mid = (l + r) >> 1
        if L <= mid:
            self._update(L, R, l, mid, root.left, delta)
        if R >= mid + 1:
            self._update(L, R, mid + 1, r, root.right, delta)
        self._pushUp(root)

    def _query(self, L: int, R: int, l: int, r: int, root: Node) -> bool:
        if L <= l <= r <= R:
            return root.isTracked

        self._pushDown(root)
        mid = (l + r) >> 1
        res = True
        if L <= mid:
            res = res and self._query(L, R, l, mid, root.left)
        if R >= mid + 1:
            res = res and self._query(L, R, mid + 1, r, root.right)
        return res

    def _pushUp(self, root: Node) -> None:
        root.isTracked = not not (
            root.left and root.left.isTracked and root.right and root.right.isTracked
        )

    def _pushDown(self, root: Node) -> None:
        if not root.left:
            root.left = Node()
        if not root.right:
            root.right = Node()
        if root.lazy:
            root.left.lazy = root.right.lazy = True
            root.left.isTracked = root.right.isTracked = root.isTracked
            root.lazy = False


# addRange,queryRange,removeRange时间均为O(logN)，N为值域，本题为1e9
# 空间复杂度为O(QlogN)，Q为所有操作的总次数


class RangeModule:
    def __init__(self):
        self.tree = SegmentTree()

    def addRange(self, left: int, right: int) -> None:
        """添加 半开区间 [left, right)"""
        self.tree.update(left, right - 1, True)

    def queryRange(self, left: int, right: int) -> bool:
        """ 只有在当前正在跟踪区间 [left, right) 中的每一个实数时，才返回 true"""
        return self.tree.query(left, right - 1)

    def removeRange(self, left: int, right: int) -> None:
        """ 停止跟踪 半开区间 [left, right)"""
        self.tree.update(left, right - 1, False)

File: 6_tree/线段树/经典题/动态开店/715. Range 模块-动态开点线段树.ts
// 线段树解法
class SegmentTreeNode {
  left!: SegmentTreeNode
  right!: SegmentTreeNode
  isLazy = false
  lazyValue = false
  value = false
}

/**
 * @description 线段树懒更新模板
 */
class SegmentTree {
  private readonly root: SegmentTreeNode = new SegmentTreeNode()
  private readonly size: number

  constructor(size: number) {
    this.size = size
  }

  update(left: number, right: number, delta: boolean): void {
    this._update(left, right, 0, this.size, this.root, delta)
  }

  query(left: number, right: number): boolean {
    return this._query(left, right, 0, this.size, this.root)
  }

  queryAll(): boolean {
    return this.root.value
  }

  private _update(
    L: number,
    R: number,
    l: number,
    r: number,
    root: SegmentTreeNode,
    delta: boolean
  ): void {
    if (L <= l && r <= R) {
      root.value = delta
      root.lazyValue = delta
      root.isLazy = true
      return
    }

    this.pushDown(root)
    const mid = (l + r) >> 1
    if (L <= mid) this._update(L, R, l, mid, root.left, delta)
    if (mid < R) this._update(L, R, mid + 1, r, root.right, delta)
    this.pushUp(root)
  }

  private _query(L: number, R: number, l: number, r: number, root: SegmentTreeNode): boolean {
    if (L <= l && r <= R) {
      return root.value
    }

    this.pushDown(root)
    let res = true
    const mid = (l + r) >> 1
    if (L <= mid) res = res && this._query(L, R, l, mid, root.left)
    if (mid < R) res = res && this._query(L, R, mid + 1, r, root.right)
    return res
  }

  /**
   * @param root 向下传递懒标记和懒更新的值 `isLazy`, `lazyValue`，并用 `lazyValue` 更新子区间的值
   */
  private pushDown(root: SegmentTreeNode): void {
    !root.left && (root.left = new SegmentTreeNode())
    !root.right && (root.right = new SegmentTreeNode())
    if (root.isLazy) {
      root.left.isLazy = true
      root.right.isLazy = true
      root.left.lazyValue = root.lazyValue
      root.right.lazyValue = root.lazyValue
      root.left.value = root.value
      root.right.value = root.value

      root.isLazy = false
      root.lazyValue = false
    }
  }

  /**
   * @param root 用子节点更新父节点的值
   */
  private pushUp(root: SegmentTreeNode): void {
    root.value = root.left.value && root.right.value
  }
}

class RangeModule {
  private readonly tree: SegmentTree = new SegmentTree(1e9 + 10)

  /**
   * @description 添加 半开区间 [left, right)
   */
  addRange(left: number, right: number): void {
    this.tree.update(left, right - 1, true)
  }

  /**
   * @description  只有在当前正在跟踪区间 [left, right) 中的每一个实数时，才返回 true
   */
  queryRange(left: number, right: number): boolean {
    return this.tree.query(left, right - 1)
  }

  /**
   * @description 停止跟踪 半开区间 [left, right)
   */
  removeRange(left: number, right: number): void {
    this.tree.update(left, right - 1, false)
  }
}

export {}
